sdk/identity/Azure.Identity/src/VisualStudioCodeCredential.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Core.Pipeline;
+using Microsoft.Identity.Client;
+
+namespace Azure.Identity
+{
+    /// <summary>
+    /// Enables authentication to Azure Active Directory using data from Visual Studio Code
+    /// </summary>
+    internal class VisualStudioCodeCredential : TokenCredential
+    {
+        private const string CredentialsSection = "VS Code Azure";
+        private const string ClientId = "aebc6443-996d-45c2-90f0-388ff96faa56";
+        private readonly IVisualStudioCodeAdapter _vscAdapter;
+        private readonly IFileSystemService _fileSystem;
+        private readonly CredentialPipeline _pipeline;
+        private readonly string _tenantId;
+
+        /// <summary>
+        /// Protected constructor for mocking
+        /// </summary>
+        protected VisualStudioCodeCredential() : this(default, default) {}
+
+        /// <inheritdoc />
+        public VisualStudioCodeCredential(string tenantId, TokenCredentialOptions options) : this(tenantId, options, default, default) {}
+
+        internal VisualStudioCodeCredential(string tenantId, TokenCredentialOptions options, IFileSystemService fileSystem, IVisualStudioCodeAdapter vscAdapter)
+        {
+            _tenantId = tenantId ?? "common";
+            _pipeline = CredentialPipeline.GetInstance(options);
+            _fileSystem = fileSystem ?? FileSystemService.Default;
+            _vscAdapter = vscAdapter ?? GetVscAdapter();
+        }
+
+        /// <inheritdoc />
+        public override async ValueTask<AccessToken> GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken)
+            => await GetTokenImplAsync(requestContext, true, cancellationToken).ConfigureAwait(false);
+
+        /// <inheritdoc />
+        public override AccessToken GetToken(TokenRequestContext requestContext, CancellationToken cancellationToken)
+            => GetTokenImplAsync(requestContext, false, cancellationToken).EnsureCompleted();
+
+        private async ValueTask<AccessToken> GetTokenImplAsync(TokenRequestContext requestContext, bool async, CancellationToken cancellationToken)
+        {
+            using CredentialDiagnosticScope scope = _pipeline.StartGetTokenScope("VisualStudioCodeCredential.GetToken", requestContext);
+
+            try
+            {
+                GetUserSettings(out var tenant, out var environmentName);
+
+                var cloudInstance = GetAzureCloudInstance(environmentName);
+                var storedCredentials = _vscAdapter.GetCredentials(CredentialsSection, environmentName);
+
+                try
+                {
+                    JsonDocument.Parse(storedCredentials);
```

Comment:
While the VSCode plugin does this check, I think all we need to do is validate that the value we got back could be a refresh token. The refresh token should be a Base64Url encoded byte array so we can just check that it only contains those characters a-z, A-Z, 0-9, _ and -

Reply:
Fixed